scipy/stats/stats.py

```diff
@@ -5718,3 +5720,175 @@ def rankdata(a, method='average'):
 
     # average method
     return .5 * (count[dense] + count[dense - 1] + 1)
+
+
+def ratio_unif(f, size=1, x1=None, x2=None, y2=None, shift=0, maxiter=None,
+               bounds=None):
+    """
+    Generate random variables using ratio-of-uniforms method.
+
+    Given a continuous univariate density f and a constant c, define the set
+    A(c) := {(x, y) : 0 < y <= sqrt(f(x/y + c)) }. If (U, V) is a random
+    vector uniformly distributed over A(c), then U/V + c follows a distribution
+    with density f.
+
+    The above result (see [1]_, [2]_) can be used to sample random variables
+    using only the density. Typical choices of c are zero or the mode of f.
+    The set A(c) is a subset of the rectangle R = [-x1, x2] x [0, y2] where
+    - x1 = inf (x - c) sqrt(f(x))
+    - x2 = sup (x - c) sqrt(f(x))
+    - y2 = sup sqrt(f(x))
+
+    In particular, these values are finite if f is bounded and x**2 * f(x) is
+    bounded (i.e. subquadratic tails). One can generate (U, V) uniformly on
+    R and return U/V + c if (U, V) are also in A(c) which can be directly
+    verified. Intuitively, this works well if A(c) fills up most of the
+    enclosing rectangle such that the probability is high that (U, V) lies in
+    A(c) whenever it lies in R(c) as the number of required iterations
+    becomes too large otherwise. To be more precise, note that the expected
+    number of iterations to draw (U, V) uniformly distributed on R(c) such that
+    (U, V) is also in A(c) is given by the ratio
+    area(R(c)) / area(A(c)) = 2 * y2 * (x2 - x1), using the fact that
+    the area of A(c) is equal to 1/2 (Theorem 7.1 in [1]_).
+
+    Parameters
+    ----------
+    f : callable
+        Scalar function defining the density.
+    size : int, optional
+        The number of random variables to generate.
+    x1 : float, optional
+        The lower bound of the bounding rectangle in the x-direction.
+        Default is None. In that case, attempt to find bound numerically.
+    x2 : float, optional
+        The upper bound of the bounding rectangle in the x-direction.
+        Default is None. In that case, attempt to find bound numerically.
+    y2 : float, optional
+        The upper bound of the bounding rectangle in the y-direction.
+        Default is None. In that case, attempt to find bound numerically.
+    maxiter : int, optional
+        Maximum number of iterations to generate the random variables.
+        Default is None. In that case, maxiter is set to a value that
+        ensures successful simulation with a high probability.
+    bounds : tuple of floats (a, b), optional
```

Comment:
It's not clear what `bounds` refers to. Does this mean bounds in `x` or in `y`? Or does it refer to vertices?

Reply:
How is this argument different to x1, x2, y2? Where is y1, is it taken to be equal to y2?

Reply:
I changed the docstring, hopefully it is more clear now. y1 is always zero, I used y2 to indicate it is also the upper bound of the rectable like x2. Maybe there is a better choice...

Reply:
Is it necessary to specify x1/x2 *and* bounds?

Reply:
no, bounds are only required if any of the Parameters x1, x2, y2 is not specified. clarified it in the doc string